src/preview/topics.rs

```diff
@@ -0,0 +1,292 @@
+use momento_protos::cache_client::pubsub::{
+    self, pubsub_client::PubsubClient, PublishRequest, SubscriptionRequest,
+};
+use tonic::{codegen::InterceptedService, transport::Channel};
+
+use crate::MomentoError;
+use crate::{
+    endpoint_resolver::MomentoEndpointsResolver,
+    grpc::header_interceptor::HeaderInterceptor,
+    utils::{connect_channel_lazily, user_agent},
+    MomentoResult,
+};
+
+type ChannelType = InterceptedService<Channel, HeaderInterceptor>;
+
+pub struct Pubsub {
+    client: pubsub::pubsub_client::PubsubClient<ChannelType>,
+}
+
+/// Work with topics, publishing and subscribing.
+impl Pubsub {
+    pub fn connect(
+        auth_token: String,
+        momento_endpoint: Option<String>,
+        user_application_name: Option<&str>,
+    ) -> MomentoResult<Self> {
+        let momento_endpoints = MomentoEndpointsResolver::resolve(&auth_token, momento_endpoint)?;
+        let channel = connect_channel_lazily(&momento_endpoints.data_endpoint.url)?;
+        let authorized_channel = InterceptedService::new(
+            channel,
+            HeaderInterceptor::new(
+                &auth_token,
+                &user_agent(user_application_name.unwrap_or("sdk")),
+            ),
+        );
+        Ok(Self {
+            client: pubsub::pubsub_client::PubsubClient::new(authorized_channel),
+        })
+    }
+
+    /// Publish a value to a topic.
+    /// The cache is used as a namespace for your topics, and it needs to exist.
+    /// You don't create topics, you just start using them.
+    pub async fn publish(
+        &self,
+        cache_name: String,
+        topic: String,
+        value: impl IntoTopicValue,
+    ) -> Result<(), MomentoError> {
+        Pubsub::actually_publish(&mut self.client.clone(), cache_name, topic, value).await
+    }
+
+    /// Publish a value to a topic.
+    /// The cache is used as a namespace for your topics, and it needs to exist.
+    /// You don't create topics, you just start using them.
+    ///
+    /// Use this if you have &mut, as it will save you a small amount of overhead for reusing the client.
+    pub async fn publish_mut(
+        &mut self,
+        cache_name: String,
```

Comment:
Ellery and I had been toying with the idea of calling this "namespace" instead of "cache" in the signatures for now until we can have a more thorough UX discussion.  Underlying implementation would be the same.  Just thinking about how thoroughly we want to permeate this choice throughout the APIs and codebases before we have actually had time to allow people to weigh in on UX.

Reply:
I'm hesitant to name the variable other than cache_name at the moment, as it definitely requires that named cache. Calling it namespace in isolation here, without also documenting somewhere appropriate "what is a namespace" and educating our customers about a whole new abstract noun's meaning and semantics I think only serves to confuse this preview implementation.

I think calling it `namespace` now would be a mistake, but I think documenting it in signatures like "this is used as a topic namespace" makes good sense. I already wrote some copy for the `momento` cli to that effect (pr blocked on this one), to help ease folks into the word.